src/ray/core_worker/core_worker.cc

```diff
@@ -400,20 +391,31 @@ Status CoreWorker::Put(const RayObject &object,
                                 static_cast<uint8_t>(TaskTransportType::RAYLET));
   reference_counter_->AddOwnedObject(*object_id, contained_object_ids, GetCallerId(),
                                      rpc_address_);
-  RAY_RETURN_NOT_OK(Put(object, contained_object_ids, *object_id));
-  // Tell the raylet to pin the object **after** it is created.
-  RAY_CHECK_OK(local_raylet_client_->PinObjectIDs(rpc_address_, {*object_id}));
-  return Status::OK();
+  return Put(object, contained_object_ids, *object_id, /*pin_object=*/true);
 }
 
 Status CoreWorker::Put(const RayObject &object,
                        const std::vector<ObjectID> &contained_object_ids,
-                       const ObjectID &object_id) {
-  RAY_CHECK(object_id.GetTransportType() ==
-            static_cast<uint8_t>(TaskTransportType::RAYLET))
-      << "Invalid transport type flag in object ID: " << object_id.GetTransportType();
-  // TODO(edoakes,swang): add contained object IDs to the reference counter.
-  return plasma_store_provider_->Put(object, object_id, nullptr);
+                       const ObjectID &object_id, bool pin_object) {
+  bool object_exists;
+  RAY_RETURN_NOT_OK(plasma_store_provider_->Put(object, object_id, &object_exists));
+  if (pin_object && !object_exists) {
+    // Tell the raylet to pin the object **after** it is created.
```

Comment:
This block seems complicated enough to move out into a separate method.

Reply:
I considered that, but there's actually not much happening here - just a single RPC call with a callback that releases the object - and moving it to a separate method would require either splitting up the release logic in sort of an ugly way or passing in a callback to that method in which case it would basically look the same.

Reply:
Oh I was thinking of moving out the whole if-else block (everything except `!object_exists`), but not a big deal.